src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -416,25 +382,41 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
 
                     int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
                     int encryptedBytes;
-                    SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
-                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    byte[] outBuffer = s_PinnableWriteBufferCache.AllocateBuffer();
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
                     {
-                        // Re-handshake status is not supported.
-                        ProtocolToken message = new ProtocolToken(null, status);
-                        throw new IOException(SR.net_io_encrypt, message.GetException());
+                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Using Pinnable", this.GetHashCode(), count, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
                     }
 
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    try
+                    {
+                        SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
+
+                        if (status.ErrorCode == SecurityStatusPalErrorCode.ContinueNeeded)
+                        {
+                            chunkBytes = 0;
+                        }
+                        else if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                        {
+                            // Re-handshake status is not supported.
+                            ProtocolToken message = new ProtocolToken(null, status);
+                            throw new IOException(SR.net_io_encrypt, message.GetException());
+                        }
+
+                    }
+                    catch
                     {
-                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Got Encrypted Buffer",
-                            this.GetHashCode(), encryptedBytes, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
+                        FreeBuffer(outBuffer);
+                        throw;
                     }
 
+
                     if (asyncRequest != null)
                     {
                         // Prepare for the next request.
                         asyncRequest.SetNextRequest(buffer, offset + chunkBytes, count - chunkBytes, s_resumeAsyncWriteCallback);
-                        Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes);
+                        Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes)
+                            .ContinueWith(_freeBufferAction, outBuffer);
```

Comment:
well for my next trick. I wanted to either a. rip out the APM, or add more state, the reason being that if we pin the input for all loops of the chunking we get more perf, but it needs more state to go with the APM calls. But I wanted buffers before bothering because pooled buffers would work better (hence my when are buffers coming questions in the other issue).

I will make the changes